ELP camera (on Jetson TX2)

The [installation documentation](https://docs.photonvision.org/en/latest/docs/getting-started/installation/coprocessor-image.html) suggests PhotonVision will run on an NVidia Jetson, but it does not work.

First, the installation script requires `curl`, but it is not pre-installed on the Jetson. An easy workaround is to manually install it first:
```
sudo apt update
sudo apt install curl
```

Once PhotonVision is running successfully, it is not able to use the attached USB camera. The following output prints repeatedly:
```
CS: WARNING: vi-output,_ov5693_2-0036: invalid JPEG image received from camera (UsbCameraImpl.cpp:528)
```

I attempted this with a Jetson TX2 using a USB ELP camera.

I think you're onto something. It looks like it's actually the camera onboard the Jetson TX2 that cscore doesn't like. When I try using cscore with python and that camera I get the same message.

However, it doesn't matter which camera I select, I am unable to get a stream to appear.

Here is the log output:
```
[2022-10-05 22:04:51] [General - Main] [ERROR] Failed to parse command-line options!
[2022-10-05 22:04:51] [General - Main] [INFO] Logging initialized in debug mode.
[2022-10-05 22:04:51] [General - Main] [INFO] Starting PhotonVision version v2022.2.0 on Unknown Platform. OS: Linux, Architecture: aarch64
[2022-10-05 22:04:51] [General - Main] [INFO] Native libraries loaded.
[2022-10-05 22:04:51] [General - ConfigManager] [INFO] Loading settings...
[2022-10-05 22:04:51] [General - ConfigManager] [INFO] Hardware config does not exist! Loading defaults
[2022-10-05 22:04:51] [General - ConfigManager] [INFO] Hardware settings does not exist! Loading defaults
[2022-10-05 22:04:51] [General - ConfigManager] [INFO] Network config file does not exist! Loading defaults
[2022-10-05 22:04:51] [General - ConfigManager] [DEBUG] Cameras config folder did not exist. Created!
[2022-10-05 22:04:51] [General - NetworkTablesManager] [INFO] Starting NT Client
[2022-10-05 22:04:51] [General - NetworkTablesManager] [ERROR] [NetworkTablesManager] Could not connect to the robot! Will retry in the background...
[2022-10-05 22:04:52] [Camera - VisionSourceManager] [DEBUG] Matching 3 new cameras!
[2022-10-05 22:04:52] [Camera - VisionSourceManager] [INFO] Adding local video device - "H264 USB Camera" at "/dev/video0"
[2022-10-05 22:04:52] [Camera - VisionSourceManager] [INFO] Adding local video device - "H264 USB Camera" at "/dev/video1"
[2022-10-05 22:04:52] [Camera - VisionSourceManager] [INFO] Adding local video device - "vi-output, ov5693 2-0036" at "/dev/video2"
[2022-10-05 22:04:52] [Camera - VisionSourceManager] [DEBUG] After matching loaded configs 3 cameras were unmatched.
[2022-10-05 22:04:52] [Camera - VisionSourceManager] [INFO] Creating a new camera config for camera H264_USB_Camera
[2022-10-05 22:04:52] [Camera - CameraConfiguration] [DEBUG] Creating USB camera configuration for UsbCameraH264 USB Camera (AKA H264_USB_Camera) at /dev/video0
[2022-10-05 22:04:52] [Camera - VisionSourceManager] [INFO] Creating a new camera config for camera H264_USB_Camera (1)
[2022-10-05 22:04:52] [Camera - CameraConfiguration] [DEBUG] Creating USB camera configuration for UsbCameraH264 USB Camera (AKA H264_USB_Camera (1)) at /dev/video1
[2022-10-05 22:04:52] [Camera - VisionSourceManager] [INFO] Creating a new camera config for camera vi-output,_ov5693_2-0036
[2022-10-05 22:04:52] [Camera - CameraConfiguration] [DEBUG] Creating USB camera configuration for UsbCameravi-output, ov5693 2-0036 (AKA vi-output,_ov5693_2-0036) at /dev/video2
[2022-10-05 22:04:52] [Camera - VisionSourceManager] [DEBUG] Matched or created 3 camera configs!
[2022-10-05 22:04:52] [WebServer - DataChangeService] [DEBUG] Added subscriber - Sources: DCS_WEBSOCKET, DCS_HTTP, DCS_NETWORKTABLES, DCS_VISIONMODULE, DCS_OTHER, Destinations: DCD_UI
[2022-10-05 22:04:52] [WebServer - DataChangeService] [DEBUG] Added subscriber - Sources: DCS_WEBSOCKET, Destinations: DCD_GENSETTINGS
[2022-10-05 22:04:53] [Camera - USBCameraSource - H264_USB_Camera (1)] [ERROR] Got a null video mode! Doing nothing...
[2022-10-05 22:04:53] [Camera - USBCameraSource - vi-output,_ov5693_2-0036] [ERROR] Got a null video mode! Doing nothing...
[2022-10-05 22:04:53] [Camera - VisionSourceManager] [DEBUG] Matched config for camera "USBFrameProvider - opencv_H264_USB_Camera" and loaded 0 pipelines
[2022-10-05 22:04:53] [Camera - VisionSourceManager] [DEBUG] Matched config for camera "USBFrameProvider - opencv_H264_USB_Camera (1)" and loaded 0 pipelines
[2022-10-05 22:04:53] [Camera - VisionSourceManager] [DEBUG] Matched config for camera "USBFrameProvider - opencv_vi-output,_ov5693_2-0036" and loaded 0 pipelines
[2022-10-05 22:04:53] [Camera - VisionSourceManager] [INFO] Adding 3 configs to VMM.
[2022-10-05 22:04:53] [General - NetworkTablesManager] [ERROR] NT Connection has failed! Will retry in background.
[2022-10-05 22:04:53] [WebServer - DataChangeService] [DEBUG] Added subscriber - Sources: DCS_WEBSOCKET, DCS_HTTP, DCS_NETWORKTABLES, DCS_VISIONMODULE, DCS_OTHER, Destinations: DCD_ACTIVEMODULE, DCD_ACTIVEPIPELINESETTINGS, DCD_GENSETTINGS, DCD_UI, DCD_OTHER
[2022-10-05 22:04:53] [VisionModule - VisionModule - H264_USB_Camera] [INFO] Setting pipeline to 0
[2022-10-05 22:04:53] [Camera - USBCameraSource - H264_USB_Camera] [DEBUG] Setting camera exposure to 50
[2022-10-05 22:04:53] [WebServer - DataChangeService] [DEBUG] Added subscriber - Sources: DCS_WEBSOCKET, DCS_HTTP, DCS_NETWORKTABLES, DCS_VISIONMODULE, DCS_OTHER, Destinations: DCD_ACTIVEMODULE, DCD_ACTIVEPIPELINESETTINGS, DCD_GENSETTINGS, DCD_UI, DCD_OTHER
[2022-10-05 22:04:53] [VisionModule - VisionModule - H264_USB_Camera (1)] [INFO] Setting pipeline to 0
[2022-10-05 22:04:53] [Camera - USBCameraSource - H264_USB_Camera (1)] [ERROR] Got a null video mode! Doing nothing...
[2022-10-05 22:04:53] [Camera - USBCameraSource - H264_USB_Camera (1)] [DEBUG] Setting camera exposure to 50
[2022-10-05 22:04:53] [WebServer - DataChangeService] [DEBUG] Added subscriber - Sources: DCS_WEBSOCKET, DCS_HTTP, DCS_NETWORKTABLES, DCS_VISIONMODULE, DCS_OTHER, Destinations: DCD_ACTIVEMODULE, DCD_ACTIVEPIPELINESETTINGS, DCD_GENSETTINGS, DCD_UI, DCD_OTHER
[2022-10-05 22:04:53] [VisionModule - VisionModule - vi-output,_ov5693_2-0036] [INFO] Setting pipeline to 0
[2022-10-05 22:04:53] [Camera - USBCameraSource - vi-output,_ov5693_2-0036] [ERROR] Got a null video mode! Doing nothing...
[2022-10-05 22:04:53] [Camera - USBCameraSource - vi-output,_ov5693_2-0036] [ERROR] Failed to set camera brightness!
[2022-10-05 22:04:53] [Camera - USBCameraSource - vi-output,_ov5693_2-0036] [ERROR] VideoException [edu.wpi.first.cscore.VideoException: property write failed]
        at edu.wpi.first.cscore.CameraServerJNI.setCameraBrightness(Native Method)
        at edu.wpi.first.cscore.VideoCamera.setBrightness(VideoCamera.java:27)
        at org.photonvision.vision.camera.USBCameraSource$USBCameraSettables.setBrightness(USBCameraSource.java:135)
        at org.photonvision.vision.processes.VisionModule.setPipeline(VisionModule.java:394)
        at org.photonvision.vision.processes.VisionModule.<init>(VisionModule.java:149)
        at org.photonvision.vision.processes.VisionModuleManager.addSources(VisionModuleManager.java:59)
        at org.photonvision.vision.processes.VisionSourceManager.tryMatchUSBCams(VisionSourceManager.java:83)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)

[2022-10-05 22:04:53] [Camera - USBCameraSource - vi-output,_ov5693_2-0036] [DEBUG] Setting camera exposure to 50
[2022-10-05 22:04:53] [Camera - USBCameraSource - vi-output,_ov5693_2-0036] [ERROR] Failed to set camera exposure!
[2022-10-05 22:04:53] [Camera - USBCameraSource - vi-output,_ov5693_2-0036] [ERROR] VideoException [edu.wpi.first.cscore.VideoException: property write failed]
        at edu.wpi.first.cscore.CameraServerJNI.setCameraExposureManual(Native Method)
        at edu.wpi.first.cscore.VideoCamera.setExposureManual(VideoCamera.java:74)
        at org.photonvision.vision.camera.USBCameraSource$USBCameraSettables.setExposure(USBCameraSource.java:124)
        at org.photonvision.vision.processes.VisionModule.setPipeline(VisionModule.java:395)
        at org.photonvision.vision.processes.VisionModule.<init>(VisionModule.java:149)
        at org.photonvision.vision.processes.VisionModuleManager.addSources(VisionModuleManager.java:59)
        at org.photonvision.vision.processes.VisionSourceManager.tryMatchUSBCams(VisionSourceManager.java:83)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)

[2022-10-05 22:04:53] [WebServer - SocketHandler] [INFO] New websocket connection from /192.168.1.215:52707
[2022-10-05 22:04:54] [General - HardwareManager] [INFO] Shutting down LEDs...
```

--- a/photon-core/src/main/java/org/photonvision/vision/camera/USBCameraSource.java
+++ b/photon-core/src/main/java/org/photonvision/vision/camera/USBCameraSource.java
@@ -65,7 +65,12 @@ public class USBCameraSource extends VisionSource {
             disableAutoFocus();
 
             usbCameraSettables = new USBCameraSettables(config);
-            usbFrameProvider = new USBFrameProvider(cvSink, usbCameraSettables);
+            if (usbCameraSettables.getAllVideoModes().isEmpty()) {
+                logger.info("Camera " + camera.getPath() + " has no video modes supported by PhotonVision");
+                usbFrameProvider = null;
+            } else {
+                usbFrameProvider = new USBFrameProvider(cvSink, usbCameraSettables);
+            }
         }
     }
 

--- a/photon-core/src/main/java/org/photonvision/vision/processes/VisionSourceManager.java
+++ b/photon-core/src/main/java/org/photonvision/vision/processes/VisionSourceManager.java
@@ -318,7 +318,8 @@ public class VisionSourceManager {
 
             var newCam = new USBCameraSource(configuration);
 
-            if (!newCam.cameraQuirks.hasQuirk(CameraQuirk.CompletelyBroken)) {
+            if (!newCam.cameraQuirks.hasQuirk(CameraQuirk.CompletelyBroken)
+                    && !newCam.getSettables().videoModes.isEmpty()) {
                 cameraSources.add(newCam);
             }
         }

